src/search.coffee

```diff
@@ -6,234 +6,257 @@ url = require 'url'
+        existingIds = []
+        responses.forEach (response) ->
+          data.count = response.count if response.count?
+          response.results.forEach (result) ->
+            unless result.id in existingIds
+              data.results.push result
+              existingIds.push result.id
+
+        callback null, data
+
+  counts: (query, options, callback) ->
+    [callback, options] = [options, {}] if _.compact(arguments).length is 2
+
+    sources = ['web', 'images', 'videos', 'news']
+    methods = {
+      web: _.bind @_rawWeb, this
+      images: _.bind @_rawImages, this
+      videos: _.bind @_rawVideos, this
+      news: _.bind @_rawNews, this
+    }
+
+    executeSearch = (source, callback) =>
+      # We only need to one result (zero is impossible) from most verticals.
+      # With a web search, however, the `totalEstimatedMatches` needs to be
+      # checked from a higher page for accurate data.
+      #
+      # The 1,000 value comes from empirical data. It seems that after 600
+      # results, the accuracy gets quite consistent and accurate. I picked 1,000
+      # just to be in the clear. It also doesn't matter if there are fewer than
+      # 1,000 results.
+      pagination = {offset: 0, top: 1}
+      if source is 'web'
+        _.each pagination, (value, key) ->
+          pagination[key] += 1000
+
+      methods[source] query, (_.defaults pagination, options), callback
+
+    async.mapLimit sources, @parallelLimit, executeSearch,
+      (err, responses) ->
+        callback err if err?
+
+        data = {}
+        data[source] = 0 for source in sources
+
+        responses.forEach (response, i) ->
+          data[sources[i]] = response.count
+
+        callback null, data
+
+  # This modifies the endpoint used for searching to retrieve params specific
+  # to separate verticals (i.e. width/height for images and runtime for videos).
+  _verticalSearch: (vertical, q, options, callback) ->
+    options = _.extend({}, options, {q, endpoint: "#{vertical}/search"})
+    @_parallelSearch options, callback
+
+  # This sends the vertical via the responseFilters query param for methods
+  # which don't have specific verticals (i.e. web, spelling, related).
+  _filteredSearch: (responseFilter, q, options, callback) ->
+    options = _.extend({}, options, {q, responseFilter})
+    @_parallelSearch options, callback
+
+  _rawWeb: (query, options, callback) ->
+    @_filteredSearch 'webpages', query, options, callback
 
   web: (query, options, callback) ->
-    @verticalSearch 'Web', _.bind(@extractWebResults, this), query, options,
-      callback
+    [callback, options] = [options, {}] if _.compact(arguments).length is 2
 
-  extractWebResults: (results) ->
-    @mapResults results, ({ID, Title, Description, DisplayUrl, Url}) ->
-      id: ID
-      title: Title
-      description: Description
-      displayUrl: DisplayUrl
-      url: Url
+    @_rawWeb query, options, (err, data) =>
+      callback err if err?
```

Comment:
`return` and newline after early return.

You don't need an existential operator for the err check either. If it is anything other than `null` or `undefined` (and technically `0` or `''`), it will be truthy.

Also, when the callbacks are small like this and really the only difference is whether there is an error or not, you can combine these like so:

```coffee
callback err, (@_extractWebResults data unless err)
```